File: minigit-3/py/commands/hash_object.py
"""hash-object command - Compute object hash."""

import sys
from pathlib import Path


def run(args: list[str]) -> int:
    """Compute object hash."""
    from utils import find_repo_root, hash_object_data, write_object

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    write = False
    obj_type = 'blob'
    file_path = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-w':
            write = True
        elif arg == '-t' and i + 1 < len(args):
            obj_type = args[i + 1]
            i += 1
        elif not arg.startswith('-'):
            file_path = arg
        i += 1

    if file_path is None:
        print("error: file required", file=sys.stderr)
        return 1

    # Read file
    path = Path(file_path)
    if path.is_absolute():
        full_path = path
    else:
        full_path = repo_root / path

    if not full_path.exists():
        print(f"error: file '{file_path}' not found", file=sys.stderr)
        return 1

    content = full_path.read_bytes()

    if write:
        sha = write_object(repo_root, obj_type, content)
    else:
        sha = hash_object_data(obj_type, content)

    print(sha)
    return 0

File: minigit-3/ts/src/commands/hash-object.ts
// hash-object command - Compute object hash

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, sha1 } from '../utils';
import { writeBlob, createBlobContent, hashObject, writeObject } from '../objects';

export function hashObjectCmd(args: string[]): void {
  let writeFlag = false;
  let objectType = 'blob';
  let filePath: string | null = null;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-w') {
      writeFlag = true;
    } else if (arg === '-t' && i + 1 < args.length) {
      objectType = args[++i];
    } else if (!arg.startsWith('-')) {
      filePath = arg;
    }
  }

  if (!filePath) {
    console.error('fatal: file required');
    process.exit(1);
  }

  const absPath = path.resolve(process.cwd(), filePath);

  if (!fs.existsSync(absPath)) {
    console.error(`fatal: could not open '${filePath}' for reading: No such file or directory`);
    process.exit(1);
  }

  const content = fs.readFileSync(absPath);

  // Create object content with header
  const header = `${objectType} ${content.length}\0`;
  const fullContent = Buffer.concat([Buffer.from(header), content]);
  const hash = sha1(fullContent);

  if (writeFlag) {
    const repoRoot = findRepoRoot();
    if (!repoRoot) {
      console.error('fatal: not a git repository');
      process.exit(1);
    }
    writeObject(fullContent, repoRoot);
  }

  console.log(hash);
}
